File: CPlusPlusPractice/source/ValidParentheses.cpp
// https://leetcode.com/problems/valid-parentheses
#include <string>
#include <stack>
#include <unordered_map>

class Solution
{
    std::unordered_map<char, char> bracketMap = { { ')', '(' }, { '}', '{' }, { ']', '[' } };
public:
    bool isValid(std::string s)
    {
        if (s.empty()) return true;

        std::stack<char> stack;
        for (const char c : s)
        {
            if (c == '(' || c == '{' || c == '[')
            {
                stack.push(c);
            }
            else
            {
                if (!stack.empty() && stack.top() == bracketMap[c])
                {
                    stack.pop();
                }
                else
                {
                    return false;
                }
            }
        }
        return stack.empty();
    }
};

File: CSharpPractice/source/ValidParentheses.cs
// https://leetcode.com/problems/valid-parentheses
using System.Collections.Generic;

namespace LeetCodePractice.ValidParentheses
{
    public class Solution
    {
        private Dictionary<char, char> bracketMap = new Dictionary<char, char> { { '}', '{' }, { ']', '[' }, { ')', '(' } };
        public bool IsValid(string s)
        {
            if (s.Length == 0) return true;

            Stack<char> stack = new Stack<char>();
            foreach (char c in s)
            {
                if (c == '{' || c == '[' || c == '(')
                {
                    stack.Push(c);
                }
                else
                {
                    if (stack.Count > 0 && stack.Peek() == bracketMap[c])
                    {
                        stack.Pop();
                    }
                    else
                    {
                        return false;
                    }
                }
            }
            return stack.Count == 0;
        }
    }
}